doc: small fixes in the buffer.md

vsemozhetbyt opened the pull request:
##### Checklist

- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

doc, buffer

##### Description of change

1. `var` => `const` / `let`.
2. Fix a syntax error.
3. Fix wrong arguments in function calls*.
4. Fix some copy-paste artifacts.
5. Repeat the remark about buffer module property where it is needed.
6. Add a missing variable declaration.
7. Remove / fix wrong remarks in the buffer.md
8.  Fix examples in buffer.md to avoid confusion (fixes #9786)
9.  Remove an extraneous word in the buffer.md

*As to `null` => `undefined` in point 3, it is a temporary fix untill #9801 is addressed.

sam-github commented on doc/api/buffer.md:
````diff
@@ -404,14 +404,14 @@ to initialize a `Buffer` to zeroes.
 
 ```js
-const buf = new Buffer(5);
+const buf = new Buffer(10);
````
how is this a fix? why is `10` better than `5`?

vsemozhetbyt replied:
See #9786. This increase the size of example (with comments) so the output could not be confusing.

vsemozhetbyt replied:
Sorry for my incomprehensibility. Consider this case: a user copies this example, pastes it in the empty file `test.js`, saves it and run via console or IDE to see the output. He will not see described output, he will see two identical lines with null bytes all the times. In #9876 I've described the cause. This fix increases the size of possible `test.js` so the two lines will be different. I'll try to make some GIFs now to explain more clearly what I mean.

vsemozhetbyt replied:
In the first GIF you can see a test attempt with the old code example. See how both output lines are the same in contravention of the doc prediction.

![buffer1](https://cloud.githubusercontent.com/assets/10393198/20635739/e88090ec-b369-11e6-9610-c788ce6b6174.gif)

In the second GIF you can see a test attempt with the new code example. See how both output lines are different as the doc predicts.

![buffer2](https://cloud.githubusercontent.com/assets/10393198/20635742/f5bf9442-b369-11e6-99ae-81af35839bb4.gif)

sam-github replied:
```
core/tls-node (notes-tls $% u=) % node f.js
<Buffer 48 06 65 02 00>
<Buffer 00 00 00 00 00>
core/tls-node (notes-tls $% u=) % cat f.js
const buf = new Buffer(5);
// Prints: (contents may vary): <Buffer 78 e0 82 02 01>
console.log(buf);
buf.fill(0);

// Prints: <Buffer 00 00 00 00 00>
console.log(buf);
```

Contents may vary mean it can be anything, including zeroed. Maybe just add a comment to that effect? Looks like your system might always zero buffers 8 bytes or smaller?

sam-github replied:
I guess I don't mind the buffer getting a bit bigger, but if you are hoping that means it will always be non-zero.... I don't think that's true. Your own system proves that the initial values can be anything, right? `10` bytes is not special, AFAICT.

sam-github replied:
Thus I think the docs should make the behaviour clear, not the examples. Examples show only one scenario, the docs say how it will be for *all* scenarios. And frankly, I don't get the point of this example, it doesn't add anything to my understanding. The docs already said the memory wouldn't be initialized. But that's another issue, I guess.

sam-github replied:
I don't know. It clearly confused you, and docs should clarify, not confuse. On the other hand, 10 seems as random as 5. I guess leave it, at least anyone wondering why it was changed can find the PR, this conversation, and understand why it changed.

vsemozhetbyt replied:
Thank you for the patience.

So can I remake the PR as I've described below?

sam-github commented on doc/api/buffer.md:
````diff
@@ -523,14 +523,14 @@ initialized*. The contents of the newly created `Buffer` are unknown and
 
 ```js
-const buf = Buffer.allocUnsafe(5);
+const buf = Buffer.allocUnsafe(10);
````
ditto

vsemozhetbyt replied:
ditto

sam-github commented on doc/api/buffer.md:
```diff
@@ -1264,7 +1264,7 @@ console.log(buf.lastIndexOf('buffer', 4));
 
 // Prints: 6
-console.log(utf16Buffer.lastIndexOf('\u03a3', null, 'ucs2'));
+console.log(utf16Buffer.lastIndexOf('\u03a3', buf.length, 'ucs2'));
```
`null` didn't work? It looks like the example was trying to show how the second argument has a sensible default, and you have just provided the default value explicitly.

vsemozhetbyt replied:
However, I am not so sure about this now. It seems the doc is right, but the lib code is wrong. See https://github.com/nodejs/node/blob/master/lib/buffer.js#L603. Maybe we should fix the lib instead.

vsemozhetbyt replied:
It becomes more weird: https://github.com/nodejs/node/blob/master/test/parallel/test-buffer-indexof.js#L407

vsemozhetbyt replied:
I've created [an issue](https://github.com/nodejs/node/issues/9801) concerning this.

vsemozhetbyt replied:
Maybe `null` could be changed into `undefined` for now (`undefined` does print `6` here). And if #9801 is fixed, we could restore the `null` .

vsemozhetbyt commented:
@sam-github Well, I could try, but I am a bit confused how to do this. If I add several commits to branch, will the first one remain in the history? Should I close this PR and make a new one from scratch? Sorry, I am not a git pro, I usually use GitHub interface(

vsemozhetbyt commented:
I've tried to understand the git commands for this case. Can I do `git commit --amend --all` and `git push --force-with-lease` fo my branch safely for this PR?

vsemozhetbyt commented:
So maybe I'll try it now. If this causes a mess, I'll just close the PR and recreate it with a link to this conversation.

vsemozhetbyt commented:
Does it seem good?

vsemozhetbyt commented:
@sam-github Does it seem OK now?

sam-github commented on doc/api/buffer.md:
```diff
@@ -2306,6 +2306,9 @@ added: v3.0.0
 On 32-bit architectures, this value is `(2^30)-1` (~1GB).
 On 64-bit architectures, this value is `(2^31)-1` (~2GB).
 
+Note that this is a property on the `buffer` module as returned by
```
the word "as" is extraneous

sam-github commented:
`doc: fix examples in buffer.md to avoid confusion`, the commit message is confusing, because its not obvious what is confusing about 5, that 10 fixes. The message body should say something like "On some systems, small buffer allocations can be zeroed by default, so the example doesn't work well - the buffer looks zeroed even before the fill."

vsemozhetbyt commented:
@sam-github I've removed "as" from the initial remark and from the both my repetition. Also, I've added the requested commit message body with small edition: it's not the small buffer that zeroed, it's buffer beginning, i.e. some first bytes.

vsemozhetbyt commented:
ping @nodejs/documentation

vsemozhetbyt commented:
@sam-github Should I cc somebody for another review?

sam-github commented:
I and a number of folks are at node interactive this week, I'd give it a few days before worrying that you've been forgotten. And I think two reviews are usually wanted, @nodejs/documentation, maybe someone else can review?

silverwind commented on doc/api/buffer.md:
```diff
@@ -1223,7 +1223,7 @@ added: v6.0.0
 * `value` {String | Buffer | Integer} What to search for
-* `byteOffset` {Integer} Where to begin searching in `buf` (not inclusive).
+* `byteOffset` {Integer} Where to begin searching in `buf`.
   **Default:** [`buf.length`]
```
Default is actually `buf.length - 1`.

https://github.com/nodejs/node/blob/e21e12944c11f9d30e154d735f35b1f528cd7dc2/lib/buffer.js#L594

vsemozhetbyt replied:
Should I preserve the link? Should it be
```
[`buf.length - 1`][`buf.length`]
```

vsemozhetbyt replied:
@silverwind Should I preserve the link? Should it be
```
[`buf.length - 1`][`buf.length`]
```

silverwind replied:
Put the `- 1` outside the link.

silverwind commented on doc/api/buffer.md:
```diff
@@ -1264,7 +1264,7 @@ console.log(buf.lastIndexOf('buffer', 4));
 
 // Prints: 6
-console.log(utf16Buffer.lastIndexOf('\u03a3', null, 'ucs2'));
+console.log(utf16Buffer.lastIndexOf('\u03a3', undefined, 'ucs2'));
```
Both null and undefined are technically not valid arguments.

vsemozhetbyt replied:
The same here, see https://github.com/nodejs/node/issues/9801 . Maybe this could be postponed till that big issue fixed?

vsemozhetbyt replied:
@silverwind It is a bit complicated here and I hoped it would be fixed fully when #9801 is addressed. Maybe this could be postponed till that big issue fixed?

silverwind replied:
Yeah fine with me.

vsemozhetbyt commented:
@silverwind Is it OK this way?

vsemozhetbyt commented:
@sam-github, @silverwind  Does it require more reviews to land?

silverwind commented:
Not strictly, but I'd like to give it a day or two before landing in case someone else wants to chime in.

vsemozhetbyt commented:
@sam-github , @silverwind , @addaleax It seems we have concurrent PRs with #10151 , if I get it right. Or there will be no conflict if diff fragments are not the same?

addaleax commented:
@vsemozhetbyt They seem to merge without conflicts, so, yay, everything’s fine. \o/

I’m landing this now if you don’t mind (but thanks for pointing out the possible conflict!)

vsemozhetbyt commented:
Oh, I see. Sorry for false bothering)

addaleax commented:
Landed in 8951d3e0e551...898c25aa6497, thank you!

TheAlphaNerd commented:
the number of discrete changes in this PR makes it a bit difficult to backport. Would someone be willing to manually backport
